libraries/Microsoft.Bot.Builder/AppInsightsLoggerMiddleware.cs

```diff
@@ -0,0 +1,151 @@
+        public AppInsightsLoggerMiddleware(string instrumentationKey, bool logUserName = false, bool logOriginalMessage = false, TelemetryConfiguration config = null)
+        {
+            if (instrumentationKey == null)
+            {
+                throw new ArgumentNullException(nameof(instrumentationKey));
+            }
+
+            var telemetryConfiguration = config ?? new TelemetryConfiguration(instrumentationKey);
+            _telemetryClient = new TelemetryClient(telemetryConfiguration);
+            _logUserName = logUserName;
+            _logOriginalMessage = logOriginalMessage;
+        }
+
+        /// <summary>
+        /// Records incoming and outgoing activities to the Application Insights store.
+        /// </summary>
+        /// <param name="context">The context object for this turn.</param>
+        /// <param name="nextTurn">The delegate to call to continue the bot middleware pipeline.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        /// <seealso cref="ITurnContext"/>
+        /// <seealso cref="Bot.Schema.IActivity"/>
+        public async Task OnTurnAsync(ITurnContext context, NextDelegate nextTurn, CancellationToken cancellationToken)
+        {
+
+            context.Services.Add(AppInsightServiceKey, _telemetryClient);
+
+            // log incoming activity at beginning of turn
+            if (context.Activity != null)
+            {
+                if (string.IsNullOrEmpty((string)context.Activity.From.Properties["role"]))
+                {
+                    context.Activity.From.Properties["role"] = "user";
+                }
+
+                // Context properties for App Insights
+                if (!string.IsNullOrEmpty(context.Activity.Conversation.Id))
+                {
+                    _telemetryClient.Context.Session.Id = context.Activity.Conversation.Id;
+                }
+
+                if (!string.IsNullOrEmpty(context.Activity.From.Id))
+                {
+                    _telemetryClient.Context.User.Id = context.Activity.From.Id;
+                }
+
+                // Log the Application Insights Bot Message Received
+                try
+                {
+                    _telemetryClient.TrackEvent(BotMsgEvent, FillEventProperties(context));
+                }
+                catch (Exception ex)
+                {
+                    _telemetryClient.TrackException(ex);
+                }
+            }
+
+            // process bot logic
+            await nextTurn(cancellationToken).ConfigureAwait(false);
+        }
+
+        private static Dictionary<string, string> FillEventProperties(ITurnContext context)
+        {
+            // Payload properties
+            var properties = new Dictionary<string, string>()
```

Comment:
@cleemullins should we establish a code analysis rule around either requiring or prohibiting using an empty constructor when object initialization syntax is being used? My preference is eliding them since it's just more noise, but I'll do whichever one... it's just that the code base currently contains both approaches today.

Reply:
I don't think such a rule exists today. Writing rules looks *really* easy though, although the process by which those rules are taken into the product seems very detailed.

Reply:
I'll leave it for now - if we want to make a stance on this, we can change..